### PR TITLE
[Datasets] Add FAQ to Datasets docs.

### PR DESCRIPTION
**Pull request by clarkzinzow**
This PR adds a FAQ to Datasets docs.

Docs preview: https://ray--24932.org.readthedocs.build/en/24932/

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by clarkzinzow** [comment]
@ericl @jianoaix I added a first pass at each of the FAQ, haven't proof-read yet or added some external links, but I thought I'd ping y'all for an early review.

**Review by ericl — COMMENTED** [review]
I made some minor edits; there also seem to be some broken link refs.

Main comment is that Torch / TF datasets need to be trimmed down. Keep it focused on the top 2-3 bullets to avoid overwhelming the reader.

Rest of sections look good at first pass.

**Comment by clarkzinzow** [comment]
@ericl Please don't push to this PR while I'm actively working on it, I just accidentally overwrote your changes. Can you rebase your changes and push them back up?

**Comment by ericl** [comment]
I just used the git edit feature so that's lost forever... consider not force pushing in the future.

**Comment by ericl** [comment]
Here are the edits I had previously. Most are correcting for tone suitable for technical documentation / marketing material:

> We're happy to say that we know of several Ray Datasets users that are running their
Datasets integrations in production, both via OSS Ray and Anyscale. We give a few
notable integrations below:

Rewrite for tone: "To give an idea of Datasets use cases, we list a few notable users running Dataset integrations in production below:"

> Ray Datasets has great integration
with these frameworks, allowing for efficient exchange of distributed data partitions
often involving no data movement or even data copying!

"Datasets integrates with these frameworks, allowing for efficient exchange of distributed data partitions often with zero-copy."

> Below, we briefly summarize where we think Ray Datasets is different
and/or adds value.

"Below, we summarize some advantages Datasets offers over these more specific ingest frameworks."

> Ray Datasets doesn't (yet) have a featureful query compiler, so some manual performance

"Datasets doesn't perform query optimization, so some manual performance"

**Comment by clarkzinzow** [comment]
@ericl I made those tone changes.

As for trimming down the Torch/TensorFlow datasets comparisons, do you have any opinions on which should be eliminated or merged? IMO each are important differentiators, most are pain-points we've heard from users.

File: doc/source/data/faq.rst [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+management.
+
+Who is using Ray Datasets?
+==========================
+
+We're happy to say that we know of several Ray Datasets users that are running their
+Datasets integrations in production, both via OSS Ray and Anyscale. We give a few
+notable integrations below:
+
+* Predibase is using Ray Datasets for ML ingest and batch inference in their OSS
```
[the remark this answers is not in the document]
Comment (jianoaix) [reply doc/source/data/faq.rst:44]: Ok, we shouldn't mention it as of now. (maybe we should delete this conversation as well)

File: doc/source/data/faq.rst [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+
+* `NVTabular <https://github.com/NVIDIA-Merlin/NVTabular>`__ only supports tabular
+  (Parquet, CSV, Avro) data, while Ray Datasets supports many other file formats.
+* NVTabular doesn't support mixing heterogeneous resources in dataset transforms (e.g.
+  both CPU and GPU transformations), while Ray Datasets supports this.
+* NVTabular has a bunch of great ML-specific preprocessing operations; this is current
+  WIP for Ray Datasets via the :ref:`Ray AIR preprocessors <air-key-concepts>`.
+
+For batch (offline) inference, why should I use Ray Datasets instead of an actor pool?
+======================================================================================
```
[the remark this answers is not in the document]
Comment (jianoaix) [reply doc/source/data/faq.rst:190]: That question is indeed more important, and this one is more like if user is already in Ray ecosystem, why they should use Dataset rather than DIY with Ray's lower-level APIs. For the first version, it sounds fine to drop this question.

**Comment by clarkzinzow** [comment]
@ericl I was making this kind of categorization when I saw your comment, I like your Datasets-oriented categories, so I aligned it to that. Lmk what you think.

File: doc/source/data/faq.rst [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+    :local:
+    :depth: 2
+
+
+What problems does Ray Datasets solve?
+======================================
+
+Ray Datasets aims to solve the problems of:
+
+1. Slow, unpipelined, and/or unscalable preprocessing and ingestion pipelines for
```
Comment (maxpumperla) [inline doc/source/data/faq.rst:26]: Can we find a better way to express "unpipelined"? Depending on how favourably one wants to read this, this might be confused as "unpipelined [...] ingestion pipelines".

Also, the two bullet points are almost identical. Can we do something like:

Ray Datasets aims to solve the problem dealing with with slow or unscalable processing steps and ingestion pipelines for two primary use cases:

1. For **model training**, ...
2. For ...

Comment (maxpumperla) [reply doc/source/data/faq.rst:26]: also, the way it's currently formulated does not quite make clear, what exactly the problem is.

Comment (clarkzinzow) [reply doc/source/data/faq.rst:26]: Yep I think that common extraction makes sense, or maybe making the end-problems of performance, cost, scalability more clear.

File: doc/source/data/faq.rst [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+In order to solve these problems without sacrificing usability, Ray Datasets simplifies
+parallel and pipelined data processing on Ray, providing a higher-level API while
+internally handling data batching, task parallelism and pipelining, and memory
+management.
+
+Who is using Ray Datasets?
+==========================
+
+To give an idea of Datasets use cases, we list a few notable users running Datasets
+integrations in production below:
```
Comment (maxpumperla) [inline doc/source/data/faq.rst:42]: Great, really like this one!

File: doc/source/data/faq.rst [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+and efficiency marks for these use cases.
+
+See our :ref:`ML preprocessing docs <datasets-ml-preprocessing>` for more information on
+this positioning.
+
+What should I not use Ray Datasets for?
+=======================================
+
+Ray Datasets is not meant to be used for generic ETL pipelines (like Spark) or
+scalable data science (like Dask, Modin, or Mars). Datasets integrates with these
```
Comment (maxpumperla) [inline doc/source/data/faq.rst:86]: maybe this is too much to ask here, but naively speaking what's the notable diff between data science as done by e.g. Dask and our above use case of "loading, prep. and ingesting data for ML"? What exactly are we missing in this positioning? I'd rather be very clear about this here.

File: doc/source/data/faq.rst [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+  shards. See our
+  `blog post <https://www.anyscale.com/blog/deep-dive-data-ingest-in-a-third-generation-ml-architecture>`__
+  on why this shared infrastructure is important for 3rd generation ML architectures.
+* **Lower overhead:** Datasets is lower overhead: it supports zero-copy exchange between
+  processes, in contrast to the multi-processing-based pipelines of Torch datasets.
+
+TensorFlow datasets
+~~~~~~~~~~~~~~~~~~~
+
+* **Framework-agnostic:** Datasets is framework-agnostic and portable between different
```
Comment (maxpumperla) [inline doc/source/data/faq.rst:123]: Many of these points are duplicated. Instead of grouping by competitor, list our strengths and mention how competitors are lacking in comparison.

Comment (maxpumperla) [reply doc/source/data/faq.rst:123]: @clarkzinzow not sure how you want to handle this in the end, just saw @ericl 's comment on the subject, too.

File: doc/source/data/faq.rst [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+===============================================================
+
+Pipelining is useful in a few scenarios:
+
+* You have two chained operations using different resources (e.g. CPU and GPU) that you
+  want to saturate; this is the case for both ML ingest (CPU-based preprocessing and
+  GPU-based training) and batch inference (CPU-based preprocessing and GPU-based batch
+  inference).
+* You want to do streaming data loading and processing in order to keep the size of the
+  working set small; see previous FAQ on how to do streaming data loading and
```
Comment (maxpumperla) [inline doc/source/data/faq.rst:248]: can we just link to that?

File: doc/source/data/faq.rst [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+  processing.
+* You want to decrease the time-to-first-batch (latency) for a certain operation at the
+  end of your workload. This is the case for training and inference since this prevents
+  GPUs from being idle (which is costly), and can be advantageous for some other
+  latency-sensitive consumers of datasets.
+
+When should I use global per-epoch shuffling?
```
Comment (maxpumperla) [inline doc/source/data/faq.rst:255]: Maybe quickly mention what that is and why it's important. :P

Comment (maxpumperla) [reply doc/source/data/faq.rst:255]: (the "important" part to handle below, as I see now)

File: doc/source/data/faq.rst [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+======================================================
+
+Ray Datasets doesn't perform query optimization, so some manual performance
+tuning may be necessary depending on your use case and data scale. Please see our
+:ref:`performance tuning guide <data_performance_tips>` for more information.
+
+How can I contribute to Ray Datasets?
+=====================================
+
+We're always happy to accept external contributions! If you have a question, a feature
```
Comment (maxpumperla) [inline doc/source/data/faq.rst:304]: Maybe also link to the docs contrib guide? https://docs.ray.io/en/latest/ray-contribute/docs.html

**Review by maxpumperla — COMMENTED** [review]
Looking really good! Just a couple of nits.

**Comment by clarkzinzow** [comment]
@jianoaix @ericl Ping for review

File: doc/source/data/faq.rst [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+
+Ray Datasets is not meant to be used for generic ETL pipelines (like Spark) or
+scalable data science (like Dask, Modin, or Mars). However, each of these frameworks
+are :ref:`runnable on Ray <data_integrations>`, and Datasts integrates tightly with
+these frameworks, allowing for efficient exchange of distributed data partitions often
+with zero-copy. Check out the
+:ref:`dataset creation feature guide <dataset_from_in_memory_data_distributed>` to learn
+more about these integrations.
+
+Datasets is specifically targeting
```
Comment (jianoaix) [inline doc/source/data/faq.rst:92]: nit: unintended line break?

File: doc/source/data/faq.rst [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+================================================================================
+
+There are several ML framework-specific and general solutions for loading data into
+model trainers. Below, we summarize some advantages Datasets offers over these more
+specific ingest frameworks.
+
+Torch datasets (and data loaders)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* **Framework-agnostic:** Datasets is framework-agnostic and portable between different
```
Comment (jianoaix) [inline doc/source/data/faq.rst:108]: It looks we are mostly comparing the same set of dimensions. How about groupby(comp_dimension) and then comment on each, or even better put them in a table?

**Review by jianoaix — COMMENTED** [review]
Mostly lgtm

File: doc/source/data/faq.rst (hunk 10 above)
Comment (clarkzinzow) [reply doc/source/data/faq.rst:92]: Yeah I was refactoring and didn't feel like rebalancing all lines to ~88 characters. 😛 It doesn't change the rendered text. But I can fix this while making the other changes.

File: doc/source/data/faq.rst (hunk 11 above)
Comment (clarkzinzow) [reply doc/source/data/faq.rst:108]: I was going to put that off for the sake of not having to refactor this again, but I'll take a stab at it.

Comment (clarkzinzow) [reply doc/source/data/faq.rst:108]: I have a WIP table local but I won't get it done before my laptop dies, let's do this as a follow-up.

**Comment by clarkzinzow** [comment]
@jianoaix @ericl I think that all must-have feedback has been addressed, can we merge this? I won't be able to add to this PR anymore today so I think that we should merge this and do any remaining tweaks as nice-to-have follow-ups.

**Comment by ericl** [comment]
> @jianoaix @ericl I think that all must-have feedback has been addressed, can we merge this? I won't be able to add to this PR anymore today so I think that we should merge this and do any remaining tweaks as nice-to-have follow-ups.

Merging. Please remove the author-action-required label in the future, otherwise it's easy to miss PRs that are mergeable.

**Review by ericl — APPROVED** [review]
Fixed a typo
